File: Framework.D-2015/Framework.D-2015/DigitoVerificador/DigitoVerificador.vb

Namespace DigitoVerificador

    Public Class DigitoVerificador
        Public Function calcularDV(cadena As String) As Integer

            Dim multiplo As Integer = 2
            Dim contador As Integer

            For i As Integer = cadena.Length - 1 To 0 Step -1

                Dim unChar As Char = cadena.Chars(i)
                Dim vChar As Integer = AscW(unChar)
                Dim subTot As Integer
                subTot = vChar * multiplo
                contador = contador + subTot

            Next

            contador = (11 - (contador Mod 11)) * cadena.Length

            Return contador

        End Function

        Public Function VerificarDV(cadena As String, dv As Integer) As Boolean
            If calcularDV(cadena) = dv Then
                Return True
            Else
                Return False
            End If
        End Function

    End Class
End Namespace

File: Framework.D-2015/Framework.D-2015/DigitoVerificador/DigitoVerificador.cs
using System;

namespace Framework.D_2015.DigitoVerificador
{
    public class DigitoVerificador
    {
        public int calcularDV(string cadena)
        {
            int multiplo = 2;
            var contador = 0;
            for (int i = cadena.Length - 1; i >= 0; i -= 1)
            {
                char unChar = cadena[i];
                //valor que representa el char
                int vChar = (int)unChar;
                int subTot;
                subTot = vChar * multiplo;
                contador = contador + subTot;
            }

            contador = (11 - contador % 11) * cadena.Length;
            return contador;
        }

        public bool VerificarDV(string cadena, int dv)
        {
            if (calcularDV(cadena) == dv)
            {
                return true;
            }
            else
            {
                return false;
            }
        }
    }
}